### PR TITLE
feat(marketplace): per-plugin Cowork-compatible zip export + /home Cowork UI

### PR DESCRIPTION
**Pull request by cvrysanek**
## What

Adds per-plugin **Claude Cowork** export. Cowork (`Customize → upload a custom plugin file`) is not a Claude Code marketplace consumer — its validator is stricter and expects a single plugin at the zip root. New endpoint repackages one marketplace plugin into that shape; the `/home` Cowork card and each plugin detail page expose downloads.

## Endpoint

`GET /marketplace/cowork/{prefixed_name}.zip` — same PAT/cookie auth + RBAC filtering as `marketplace.zip` (a plugin not in the caller's resolved set → 404). Cached per plugin (TTL), invalidated on marketplace sync. `CoworkZipError` → 422 when a plugin can't fit Cowork's caps.

## Transforms (matched byte-for-byte against a known-good reference upload)

Keeps **all** plugin content (`data/`, `scripts/`, `vendor/`, `global-rules/`, `CLAUDE.md`, `settings.json`, agent `tools:`). Only:
- plugin at zip root (no `marketplace.json` wrapper)
- `plugin.json` → semver `version` + required `author` + dropped `homepage`
- SKILL.md frontmatter whitelisted to `name`/`description`/`compatibility` (drops Claude-Code-only `argument-hint`/`user-invocable`); descriptions sanitized (`<`/`>` removed, `"`→`'`) and YAML-safe-encoded
- per-directory `.md` under `data/` concatenated into `_all.md` — keeps every byte while staying under Cowork's 5000-file cap (a Confluence/docs dump can be tens of thousands of files)
- Next.js route path segments renamed (`[x]`→`dyn-x`, `(y)`→`grp-y`); arcname collisions de-duplicated
- `.DS_Store` + Agnes-only paths stripped

## UI

- `/home` Cowork surface card: completed (was a placeholder), lists the analyst's RBAC-filtered plugins with per-plugin download links (from `/marketplace/info`)
- Marketplace plugin detail page: "Download for Cowork" button

## Security hardening (from an adversarial review pass)

- **Symlink/path-escape guard** — never reads a path that resolves outside the plugin dir (blocks file exfiltration via hostile plugin symlinks)
- Single-collect + cache (no repeated full-tree walk per request); total-size + file-count caps enforced
- `Content-Disposition` filename from the DB-sourced `prefixed_name` with control chars stripped (no header injection)
- YAML-safe scalar emit (a folder like `[id]` or a value like `123`/`true` stays a string, never a list/int/bool; single quotes only)

## Tests

New `tests/test_cowork_packager.py` (transforms, symlink guard, collision dedup, determinism) + `tests/test_marketplace_server_cowork.py` (auth/404/304/422). Verified the real `grpn` export is byte-identical to the reference on all 45 skill/agent files + `plugin.json`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/keboola/agnes-the-ai-analyst/pull/488" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by MonikaFeigler** [comment]
Hey @cvrysanek — thanks for this PR, the per-plugin Cowork zip export is a solid idea and the implementation is clean. Before merging, wanted to share where we are on the broader Cowork integration so we can decide together whether this is still needed.

## What we already have (PR #474 + PR #494)

We've been building Agnes Cowork integration on two fronts:

**PR #474 — Agnes MCP server** ships a Cowork setup bundle (ZIP) that wires Agnes as an MCP server in Claude Desktop. The bundle includes:
- `mcp_server.py` — pure-stdlib stdio MCP proxy, no install needed, works inside the Cowork VM on first open
- Agnes tools exposed via MCP: `catalog`, `schema`, `describe`, `query`, `pull`, `skills`
- `setup.py` — one-time token exchange that writes persistent credentials

**PR #494 (draft) — `/me/ai-connect` hub + bundle skills** extends the bundle with the thing your PR is solving — skills and agents as slash commands in Cowork. Instead of per-plugin ZIP uploads, the bundle generator now:
- Collects all RBAC-granted marketplace skills (`skills/*/SKILL.md`) and agents (`agents/*.md`) at bundle-generation time
- Writes them into `.claude/skills/` and `.claude/agents/` in the ZIP, so they appear as `/skill-name` slash commands immediately on first Cowork open
- Strips Claude-Code-only frontmatter keys (`argument-hint`, `user-invocable`) — same transform logic your `cowork_packager` does
- De-duplicates names across plugins via manifest_name prefix
- Ships three curated Agnes skills in every bundle: `/explore-data`, `/query-data`, `/new-skill`

The key UX difference vs your approach: **one bundle download → everything shows up at once**, no per-plugin upload loop. On re-open Cowork loads the same workspace with all skills already in place.

## Question before we go further

Given that the bundle already covers initial setup and all skills in one go — is the per-plugin on-demand endpoint (`GET /marketplace/cowork/{prefixed_name}.zip`) actually necessary from the user's perspective? Is there a concrete scenario where a user would need to add a single plugin to an existing Cowork session rather than just re-downloading the bundle?

Genuinely asking — if there's a real use case we're missing, it's worth keeping. But if the bundle covers it, we'd rather keep one mechanism than two.

File: app/marketplace_server/cowork_packager.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
+"""Build a single-plugin ZIP that Claude Desktop's **Cowork** upload accepts.
+
+Cowork (``Customize → upload a custom plugin file``) is NOT a Claude Code
+marketplace consumer. Its server-side validator is stricter than
+``claude plugin validate`` and expects a *single plugin* per zip, with the
+plugin at the **zip root** (no wrapper dir, no ``marketplace.json``).
+
+The transforms here are matched against a **known-good reference zip**
+(``grpn-v1.15.28.zip``) that uploaded successfully — not the (more
```
Comment (minasarustamyan) [inline app/marketplace_server/cowork_packager.py]: 🔴 **Blocker (vendor-agnostic OSS).** `grpn-v1.15.28.zip` je odkaz na reálný zákaznický artefakt (Groupon). CLAUDE.md zakazuje customer-specific tokeny v kódu/komentářích/testech/PR. Token `grpn`/`Groupon` je i v `tests/test_cowork_packager.py` (vč. `"description": "Groupon plugin"`), v `tests/test_marketplace_server_cowork.py` (fixtura `grpn`, `mkt-grpn`) a v těle PR. Přejmenovat všude na neutrální placeholder (`demo`/`sample`).

File: tests/test_cowork_packager.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    # Stripped.
+    (d / ".DS_Store").write_text("junk", encoding="utf-8")
+    # Kept root docs.
+    (d / "README.md").write_text("# grpn", encoding="utf-8")
+    (d / ".mcp.json").write_text("{}", encoding="utf-8")
+    return {
+        "manifest_name": "grpn",
+        "prefixed_name": "mkt-grpn",
+        "version": "deadbeef",
+        "raw": {"name": "grpn", "description": "Groupon plugin"},
```
Comment (minasarustamyan) [inline tests/test_cowork_packager.py]: 🔴 `"description": "Groupon plugin"` – doslova jméno zákazníka v OSS testu. Nahradit neutrálním (např. `"sample plugin"`); stejně tak `grpn`/`mkt-grpn` v celé fixtuře a v `test_marketplace_server_cowork.py`.

File: app/marketplace_server/cowork_packager.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+            while cand in seen:
+                i += 1
+                cand = f"{stem}-{i}.{ext}" if dot else f"{arc}-{i}"
+            arc = cand
+        seen.add(arc)
+        out.append((arc, data))
+    return out
+
+
+def _finalize(members: List[Tuple[str, bytes]]) -> List[Tuple[str, bytes]]:
```
Comment (minasarustamyan) [inline app/marketplace_server/cowork_packager.py:422]: 🟡 **Capy se vynucují až po načtení všeho do paměti.** `collect_members()` přečte bajty *každého* souboru do `members`, teprve pak `_finalize()` kontroluje 5000-file / 512 MB cap. Security bullet v PR prodává „total-size + file-count caps enforced" jako hardening, ale tyhle capy neomezují špičkovou paměť při sběru – přesně ten case, který PR cituje (Confluence dump o desítkách tisíc souborů), se celý načte do RAM, než ho cap odmítne. U nightly-synced obsahu tolerovatelné, ale framing to nadhodnocuje. Zvážit počítání/short-circuit přímo ve walku (count + running total).

File: app/marketplace_server/cowork_packager.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+    """
+    plugin_dir: Optional[Path] = plugin.get("plugin_dir")
+    members: List[Tuple[str, bytes]] = []
+    has_plugin_json = False
+
+    # Store-bundle entries (e.g. "flea") have no single on-disk root — their
+    # content is composed from several source dirs. Mirror packager's bundle
+    # path: merge files (minus each source's .claude-plugin/) + one synth
+    # plugin.json, then apply the same Cowork transforms.
+    if plugin.get("bundle_dirs"):
```
Comment (minasarustamyan) [inline app/marketplace_server/cowork_packager.py:452]: 🟡 **Per-file size guard chybí v bundle větvi.** Tady se `abs_path.read_bytes()` volá bez kontroly velikosti, zatímco plugin_dir větev má `if f.stat().st_size > _MAX_FILE_BYTES: continue` (ř. 453). Nekonzistentní – jeden velký soubor v bundlu se načte celý do paměti. Sjednotit kontrolu i sem.

File: app/marketplace_server/cowork_packager.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+            kept.extend(md_files)
+            continue
+        chunks = [f"# {dir_path.split('/')[-1]} — combined docs\n".encode("utf-8")]
+        for arc, data in sorted(md_files):
+            chunks.append(f"\n---\n\n## `{arc.rsplit('/', 1)[1]}`\n\n".encode("utf-8") + data)
+        kept.append((f"{dir_path}/_all.md", b"\n".join(chunks)))
+    return kept
+
+
+def _dedup_arcnames(members: List[Tuple[str, bytes]]) -> List[Tuple[str, bytes]]:
```
Comment (minasarustamyan) [inline app/marketplace_server/cowork_packager.py:374]: 🟡 **Dedup arcname může rozbít skill.** Dedup řeší kolize na úrovni *souboru*: při kolizi skill dirů (`[id]` i `dyn-id` → `skills/dyn-id/`) přejmenuje jeden `SKILL.md` na `SKILL-1.md`, čímž Cowork ten skill přestane rozpoznávat (složka musí obsahovat `SKILL.md`). Test `test_arcname_collision_deduped` ověřuje jen „žádné duplicitní arcnames", ne že oba skilly zůstanou validní. Bezpečnější by byl dedup na úrovni adresáře. Edge case, ale tichá ztráta skillu.

**Review by minasarustamyan — COMMENTED** [review]
Review – vybrané body (1× blocker, 3× should-fix). Detail v inline komentářích níže.

🔴 **Vendor-agnostic OSS** – PR zanáší zákaznicky specifické tokeny `grpn`/`Groupon` do kódu, testů, docstringu i textu PR. CLAUDE.md to má jako tvrdé pravidlo („no customer-specific content in code, comments, tests, docs … PR bodies"). Přejmenovat na neutrální (`demo`/`sample`/`acme`) na všech místech + upravit text PR.

**Comment by cvrysanek** [comment]
> Hey @cvrysanek — thanks for this PR, the per-plugin Cowork zip export is a solid idea and the implementation is clean. Before merging, wanted to share where we are on the broader Cowork integration so we can decide together whether this is still needed.
> 
> ## What we already have (PR #474 + PR #494)
> We've been building Agnes Cowork integration on two fronts:
> 
> **PR #474 — Agnes MCP server** ships a Cowork setup bundle (ZIP) that wires Agnes as an MCP server in Claude Desktop. The bundle includes:
> 
> * `mcp_server.py` — pure-stdlib stdio MCP proxy, no install needed, works inside the Cowork VM on first open
> * Agnes tools exposed via MCP: `catalog`, `schema`, `describe`, `query`, `pull`, `skills`
> * `setup.py` — one-time token exchange that writes persistent credentials
> 
> **PR #494 (draft) — `/me/ai-connect` hub + bundle skills** extends the bundle with the thing your PR is solving — skills and agents as slash commands in Cowork. Instead of per-plugin ZIP uploads, the bundle generator now:
> 
> * Collects all RBAC-granted marketplace skills (`skills/*/SKILL.md`) and agents (`agents/*.md`) at bundle-generation time
> * Writes them into `.claude/skills/` and `.claude/agents/` in the ZIP, so they appear as `/skill-name` slash commands immediately on first Cowork open
> * Strips Claude-Code-only frontmatter keys (`argument-hint`, `user-invocable`) — same transform logic your `cowork_packager` does
> * De-duplicates names across plugins via manifest_name prefix
> * Ships three curated Agnes skills in every bundle: `/explore-data`, `/query-data`, `/new-skill`
> 
> The key UX difference vs your approach: **one bundle download → everything shows up at once**, no per-plugin upload loop. On re-open Cowork loads the same workspace with all skills already in place.
> 
> ## Question before we go further
> Given that the bundle already covers initial setup and all skills in one go — is the per-plugin on-demand endpoint (`GET /marketplace/cowork/{prefixed_name}.zip`) actually necessary from the user's perspective? Is there a concrete scenario where a user would need to add a single plugin to an existing Cowork session rather than just re-downloading the bundle?
> 
> Genuinely asking — if there's a real use case we're missing, it's worth keeping. But if the bundle covers it, we'd rather keep one mechanism than two.

@MonikaFeigler we have use-case where there are not just .md, but richer content incl. tools - and then we want to upload it to cowork (not only through 1 project - but to have it generally available across different projects in cowork), plus with upload structure (not seeding as in the bundle) - there are various limits (filesize, directory structure depth, etc.) - all this facts together - your bundled setup does not work for us right now

**Review by minasarustamyan — APPROVED** [review]
